### PR TITLE
Improve do_not_merge_across_partitions_select_final optimization.

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add parallel select final for one part with level>0 when `do_not_merge_across_partitions_select_final` setting is 1. 

Relates: https://github.com/ClickHouse/ClickHouse/pull/15938 and https://github.com/ClickHouse/ClickHouse/issues/11722

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

File: tests/queries/0_stateless/01524_do_not_merge_across_partitions_select_final.sql [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
+
+INSERT INTO select_final SELECT toDate('2000-01-01'), number, '' FROM numbers(2);
+INSERT INTO select_final SELECT toDate('2000-01-01'), number, 'updated' FROM numbers(2);
+
+OPTIMIZE TABLE select_final FINAL;
+
+INSERT INTO select_final SELECT toDate('2020-01-01'), number, '' FROM numbers(2);
+INSERT INTO select_final SELECT toDate('2020-01-01'), number, 'updated' FROM numbers(2);
+
+SELECT max(x) FROM select_final FINAL where string = 'updated' SETTINGS do_not_merge_across_partitions_select_final = 1;
```
Comment (azat) [inline tests/queries/0_stateless/01524_do_not_merge_across_partitions_select_final.sql:24]: Worth checking number of threads was used (from system.query_log) ?

**Comment by azat** [comment]
>AST fuzzer (UBSan) — ../src/Functions/FunctionDateOrDateTimeAddInterval.h:92:26: runtime error: signed integer overflow: -9223372036854775807 * 60 cannot be repr

#19466

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -1305,6 +1305,12 @@ QueryPlanPtr MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
         data_settings->index_granularity,
         index_granularity_bytes);
 
+    const size_t min_marks_for_concurrent_read = roundRowsOrBytesToMarks(
+        settings.merge_tree_min_rows_for_concurrent_read,
+        settings.merge_tree_min_bytes_for_concurrent_read,
+        data_settings->index_granularity,
+        index_granularity_bytes);
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1312]:
```suggestion
    const size_t min_marks_for_concurrent_read = minMarksForConcurrentRead(
        settings.merge_tree_min_rows_for_concurrent_read,
        settings.merge_tree_min_bytes_for_concurrent_read,
        data_settings->index_granularity,
        index_granularity_bytes,
        sum_marks);
```

#19528

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 3]
```diff
@@ -1347,25 +1353,60 @@ QueryPlanPtr MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
...
                     query_info.prewhere_info,
                     true,
-                    reader_settings,
-                    virt_columns,
-                    part_it->part_index_in_query);
+                    column_names,
+                    MergeTreeReadPool::BackoffSettings(settings),
+                    settings.preferred_block_size_bytes,
+                    false);
+
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1376]: Worth adding the same message as in https://github.com/ClickHouse/ClickHouse/blob/cc11e7c2f2ea0c508c30748623d51c176a5ad21b/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L987 ?

And also count "approx" rows?

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 4]
```diff
@@ -1347,25 +1353,60 @@ QueryPlanPtr MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
...
-            for (auto part_it = parts_to_merge_ranges[range_index]; part_it != parts_to_merge_ranges[range_index + 1]; ++part_it)
+            /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
+            /// with level > 0 then we won't postprocess this part and if num_streams > 1 we
+            /// can use parallel select on this part.
+            if (num_streams > 1 && settings.do_not_merge_across_partitions_select_final &&
+                std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1 &&
+                parts_to_merge_ranges[range_index]->data_part->info.level > 0)
             {
-                auto source_processor = std::make_shared<MergeTreeSelectProcessor>(
+                MergeTreeReadPoolPtr pool = std::make_shared<MergeTreeReadPool>(
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1363]: Also shouldn't it use one pool for all parts?

**Comment by Avogar** [comment]
> If I understand correctly, then `length(thread_ids) > 1` is not correct, because there will be threads at least for:
> 
> * TCPHandler
> * PipelineExecutor
> * parts loading - https://github.com/ClickHouse/ClickHouse/blob/f996b8433499cbbabc32febf55ce2291afbc2290/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L668

Yes, you are right, I didn't think about it, thanks.

**Comment by Avogar** [comment]
@alexey-milovidov All checks passed, can I merge it?

**Comment by alexey-milovidov** [comment]
Yes, sure.

**Pull request merged**